shimmingtoolbox/coils/siemens_basis.py

```diff
@@ -0,0 +1,143 @@
+        reordered[:, :, :, 4] = spher_harm[:, :, :, 6]
+        reordered[:, :, :, 5] = spher_harm[:, :, :, 4]
+        reordered[:, :, :, 6] = spher_harm[:, :, :, 7]
+        reordered[:, :, :, 7] = spher_harm[:, :, :, 3]
+
+        return reordered
+
+    def get_scaling_factors():
+        """
+        Returns a vector of `scaling_factors` to apply to the(Siemens-reordered) 1st + 2nd order spherical harmonic fields
+        for rescaling field terms as "shim reference maps" in units of Hz/unit-shim:
+
+        Gx, Gy, and Gz should yield 1 micro-T of field shift per metre: equivalently, 0.042576 Hz/mm
+
+        2nd order terms should yield 1 micro-T of field shift per metre-squared: equivalently, 0.000042576 Hz/mm^2
+
+        Gist: given the stated nominal values, we can pick several arbitrary reference positions around the origin/isocenter
+        at which we know what the field *should* be, and use that to calculate the appropriate scaling factor.
+
+        NOTE: The method has been worked out empirically and has only been tested for 2 Siemens Prisma systems.
+        E.g.re: Y, Z, ZX, and XY terms, it was noted that their polarity had to be flipped relative to the form given by
+        shimmingtoolbox.coils.spherical_harmonics. To adapt the code to other systems, arbitrary(?) changes along these
+        lines will likely be needed.
+        """
+
+        [x_iso, y_iso, z_iso] = np.meshgrid(np.array(range(-1, 2)), np.array(range(-1, 2)), np.array(range(-1, 2)),
+                                            indexing='xy')
+        orders = np.array(range(1, 3))
+        sh = spherical_harmonics(orders, x_iso, y_iso, z_iso)
+        sh = reorder_to_siemens(sh)
+
+        n_channels = sh.shape[3]
+        scaling_factors = np.zeros(n_channels)
+
+        # indices of reference positions for normalization:
+        i_x1 = np.nonzero((x_iso == 1) & (y_iso == 0) & (z_iso == 0))
+        i_y1 = np.nonzero((x_iso == 0) & (y_iso == 1) & (z_iso == 0))
+        i_z1 = np.nonzero((x_iso == 0) & (y_iso == 0) & (z_iso == 1))
+
+        i_x1z1 = np.nonzero((x_iso == 1) & (y_iso == 0) & (z_iso == 1))
+        i_y1z1 = np.nonzero((x_iso == 0) & (y_iso == 1) & (z_iso == 1))
+        i_x1y1 = np.nonzero((x_iso == 1) & (y_iso == 1) & (z_iso == 0))
+
+        # order the reference indices like the sh field terms
+        i_ref = [i_x1, i_y1, i_z1, i_z1, i_x1z1, i_y1z1, i_x1, i_x1y1]
+
+        # distance from iso/origin to adopted reference point[units: mm]
+        r = [1, 1, 1, 1, np.sqrt(2), np.sqrt(2), 1, np.sqrt(2)]
+
+        # invert polarity
+        # Y, Z, ZX, and XY terms only(determined empirically)
+        sh[:, :, :, [1, 2, 4, 7]] = -sh[:, :, :, [1, 2, 4, 7]]
+
+        # scaling:
+        orders = [1, 1, 1, 2, 2, 2, 2, 2]
+
+        for i_ch in range(0, n_channels):
+            field = sh[:, :, :, i_ch]
+            scaling_factors[i_ch] = 42.576 * ((r[i_ch] * 0.001) ** orders[i_ch]) / field[i_ref[i_ch]]
```

Comment:
I noticed

> Gx, Gy, and Gz should yield 1 micro-T of field shift per metre: equivalently, 0.042576 Hz/mm
> 2nd order terms should yield 1 micro-T of field shift per metre-squared: equivalently, 0.000042576 Hz/mm^2

in the docstring. Which of these numbers is referred to by `42.576` in this line? Perhaps you could include a comment mentioning which value in the docstring is relevant here? My misunderstanding is likely because I'm new to the MRI world, so disregard if this is obvious to someone who has more background 😅

Reply:
It is the gyromagnetic ratio (γ = 42.576 Mhz/T) in spin angular precession: ω = γB.

Reply:
A good practice would be to include the constants in a constant variable. Since the gyromagnetic ratio is constant, you could include it in a constant named GYROMAGNETIC_RATIO = 42.576 #MHz/T. Like this it would be easier for everyone even with less knowledge in MRI. Same thing for 0.001 which is a scaling factor for units. The only constants that should be used outside of constant variables are 0 and 1.